src/api/cmd/chart-repo-sync/sync.go

```diff
@@ -0,0 +1,335 @@
+		return nil, err
+	}
+	res, err := netClient.Do(req)
+	if err != nil {
+		log.WithFields(log.Fields{"url": repoURL}).WithError(err).Error("error requesting repo index")
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		log.WithFields(log.Fields{"url": repoURL}).WithError(err).Error("error requesting repo index, are you sure this is a chart repository?")
+		return nil, errors.New("repo index request failed")
+	}
+
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	return parseRepoIndex(body)
+}
+
+func parseRepoIndex(body []byte) (*helmrepo.IndexFile, error) {
+	var index helmrepo.IndexFile
+	err := yaml.Unmarshal(body, &index)
+	if err != nil {
+		return nil, err
+	}
+	index.SortEntries()
+	return &index, nil
+}
+
+func chartsFromIndex(index *helmrepo.IndexFile, r repo) []chart {
+	var charts []chart
+	for _, entry := range index.Entries {
+		if entry[0].GetDeprecated() {
+			log.WithFields(log.Fields{"name": entry[0].GetName()}).Info("skipping deprecated chart")
+			continue
+		}
+		charts = append(charts, newChart(entry, r))
+	}
+	return charts
+}
+
+// Takes an entry from the index and constructs a database representation of the
+// object.
+func newChart(entry helmrepo.ChartVersions, r repo) chart {
+	var c chart
+	copier.Copy(&c, entry[0])
+	copier.Copy(&c.ChartVersions, entry)
+	c.Repo = r
+	c.ID = fmt.Sprintf("%s/%s", r.Name, c.Name)
+	return c
+}
+
+func importCharts(charts []chart) error {
+	var pairs []interface{}
+	var chartIDs []string
+	for _, c := range charts {
+		chartIDs = append(chartIDs, c.ID)
+		// charts to upsert - pair of selector, chart
+		pairs = append(pairs, bson.M{"_id": c.ID}, c)
+	}
+
+	db, closer := dbSession.DB()
+	defer closer()
+	bulk := db.C(chartCollection).Bulk()
+
+	// Upsert pairs of selectors, charts
+	bulk.Upsert(pairs...)
+
+	// Remove	charts no longer existing in index
+	bulk.RemoveAll(bson.M{
+		"_id": bson.M{
+			"$nin": chartIDs,
+		},
+		"repo.name": charts[0].Repo.Name,
+	})
+
+	_, err := bulk.Run()
+	return err
+}
+
+func fetchAndImportIcon(c chart) error {
+	if c.Icon == "" {
+		log.WithFields(log.Fields{"name": c.Name}).Info("icon not found")
+		return nil
+	}
+
+	req, err := http.NewRequest("GET", c.Icon, nil)
+	req.Header.Set("User-Agent", userAgent)
+	if err != nil {
+		return err
+	}
+
+	res, err := netClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("%d %s", res.StatusCode, c.Icon)
+	}
+
+	orig, err := imaging.Decode(res.Body)
+	if err != nil {
+		return err
+	}
+
+	// TODO: make this configurable?
+	icon := imaging.Fit(orig, 160, 160, imaging.Lanczos)
+
+	var b bytes.Buffer
+	imaging.Encode(&b, icon, imaging.PNG)
+
+	db, closer := dbSession.DB()
+	defer closer()
+	return db.C(chartCollection).UpdateId(c.ID, bson.M{"$set": bson.M{"raw_icon": b.Bytes()}})
+}
+
+func fetchAndImportReadmes(c chart) {
```

Comment:
this needs to be improved, we're doing it sequentially for now. However, given the order of how we import things has now changed, I think this is okay.

Reply:
Doesn't the old importer executes with a pool of x parallel workers? Could we implement that here? It seems that not much has changed in the process that will prevent us to have that importation boost from day one?

Reply:
Well, yes, but I wanted to get a first cut of this and decided to avoid implementing the worker pool straight away. How about doing it in a separate PR after this? This won't be put into use right away (the API has not been updated to serve from the database yet), so in terms of making the review process easier we can do it in a follow-up PR. Just FYI, I found that the old importer creates n goroutines (when n = number of charts) and then a buffered channel is setup as a semaphore to only allow x (some number of) goroutines to execute at a time. I think we should rework this so that we only create x goroutines and process it like a worker pool e.g. see https://gobyexample.com/worker-pools